### ISSUE TITLE
German Voice for RHVoice

### ISSUE THREAD

**Issue by winman3000** [Olga-Yakovleva/RHVoice#24]
It would be nice to have RH Voice in German. What do you Need for German language?

**Comment by abitrolly** [comment]
Training data: text, audio files that read that text, and mapping that maps audio to the text.

**Comment by winman3000** [comment]
If I understand you right, you need a nativ speaker who reads a text an 
an audio file, right?

And what you mean with:

> [...] and mapping that maps audio to the text.


You need the spoken text as a text file?

What criteria should have the read text?

Is it necessary that this text is recorded in a studio or could it be 
with a headset too?

Sorry for the beginner questions, but I am a beginner in this aria

**Comment by abitrolly** [comment]
> If I understand you right, you need a nativ speaker who reads a text an
an audio file, right?

Yes.

> You need the spoken text as a text file?

Yes.

> What criteria should have the read text?

That depends on a language. Basically the text should be chosen in a way that audio for it contains all possible combinations of sounds, or at least cover most popular.

> Is it necessary that this text is recorded in a studio or could it be
with a headset too?

It is better to record in studio with highest quality, because then you can convert audio to different formats. But.. it is possible to record with headset too.

Mind you that software is dump - it doesn't know where words starts in your audio, so you will have to mark audio files with text.

**Comment by winman3000** [comment]
OK, now I am closer to understand you. :)

> Mind you that software is dump - it doesn't know where words starts in 
> your audio, so you will have to mark audio files with text.

How can I do this? I am a blind person. If I give you a spoken text and 
the text file, whould it be possible that you can create the voice? I am 
a beginner so I am not able to create a German Voice database.

If it is too difficult to create a German Voice Database, we need 
someone who can create such database. I am not a right person to create 
such one I think.

**Comment by abitrolly** [comment]
@winman3000 how do you use RHVoice primarily? Maybe there is already a voice for your platform or there might be available training bases. I think that any training material for HTS-type synthesizer will do.

And answering your question, me personally is unlikely to create the voice, because this project is not funded and I am afraid there is still a lot of missing bits to fill in that take time. But if we get training data, it is at least possible to see what is next.

**Comment by winman3000** [comment]
> how do you use RHVoice primarily?

Currently I don't use RHVoice as it is not in German, but I've tested 
the English ones with NVDA. If a German Voice would be available in the 
future, I'll use this Voice.

> Maybe there is already a voice for your platform or there might be 
> available training bases.

There should be training bases, as MaryTTS is using HTS based voices and 
MaryTTS has German voices. But I am not able to find the training data. 
If I could find the training data, it would be easier to find the next 
steps I think. As far as I know, the German voices are publiced as a BSD 
license, so I'm afraid that we cannot use these voices. It would be nice 
to have the training materials.

**Comment by abitrolly** [comment]
Yes, MaryTTS is what I thought of too. They also provide a nice diagram of
the whole process, but details are still a bit elusive. Guys are also very
responsive. So we should ask them. I think it is also possible to hire
Olga, but I don't have contact with her.

BSD license is not a problem. It is not hard to give proper credit in voice
documentation. But we need better docs - targeted at simple users who are
not linguists or mathematicians, but who want to hack on voice
technologies. And I personally think that blind users should take the lead
and explain to everybody else how it works and what should we do. =)

**Comment by winman3000** [comment]
Have you an idea how to contact the developers of Mary TTS? I know there 
are in GitHub, but I don't know the way to contact the team leaders. The 
only thing I can do is only open a ticket.

**Comment by nshmyrev** [comment]
It's better to give a link to https://github.com/marytts/marytts/issues/440

You can also contact marytts-developers on the mailing list http://www.dfki.de/mailman/cgi-bin/listinfo/mary-users

Voice building for rhvoice is not trivial, you could probably better just use nvda plugin for openmary with openmary voices. Openmary also has better synthesis quality due to mixed excitation vocoder and advanced NLP components.

**Comment by winman3000** [comment]
I've got a link from a developer or so for German. Can you work with it?

http://www.voxforge.org/de/Downloads

**Comment by winman3000** [comment]
The bad thing in MaryTTS is that the responsiveness is not fast as in RH 
Voice. The second thing is that you need Java.

**Comment by abitrolly** [comment]
As for Voxforge files, I need to check if audio files are annotated, and if
RHVoice training pipeline can handle this format.

**Comment by abitrolly** [comment]
Right. RHVoice is optimized for responsiveness. But I don't understand why
building voice for RHVoice should be harder than for any other HTS
synthesizer. For now the main problem is the lack of complete picture.

**Comment by winman3000** [comment]
> As for Voxforge files, I need to check if audio files are annotated,

As far as I know, these files are annotated.

> and if
> RHVoice training pipeline can handle this format.

It would be very great when you can check this.

Thank you very much for your help!

**Comment by abitrolly** [comment]
Some analysis. `Prompts.tgz` contains text that should be dictated. Filename `master_prompts_train_16kHz-16bit` contains strings like:
```
ralfherzog-20080131-de71/mfc/de71-62 DIE AUSGABEN KONNTEN GESPART WERDEN
ralfherzog-20080131-de71/mfc/de71-63 MAN WIRD AUF DEN NÄCHSTEN ABSCHWUNG WARTEN MÜSSEN
ralfherzog-20080131-de71/mfc/de71-64 DA MUSS MAN AUF ANDERE EREIGNISSE WARTEN
```

I am not really sure what this label means `ralfherzog-20080131-de71/mfc/de71-63`. ralfherzog looks like a name of text and everything else is still a mystery.

**Comment by abitrolly** [comment]
Okay. http://www.repository.voxforge1.org/downloads/de/Trunk/Audio/Original/48kHz_16bit/ contains `ralfherzog-20080131-de71.tgz` Download is very slow, so I don't yet see what's inside. Looks like it should be audio for the text and name is just identifier `what-when-shortid`. `mfc/de71-64` is still unclear. 64 looks like line number, de71 a short text identifier, but what `mfc` is - it is not clear.

**Comment by winman3000** [comment]
Hmm. So we cannot build a voice for RHVoice with this stuff? ralfherzog 
is the name, but I don't know the rest. After this it is the sentence.

**Comment by nshmyrev** [comment]
I would not be that pessimistic and try well documented openmary voice import procedure first.

**Comment by abitrolly** [comment]
@nshmyrev is that it? http://mary.opendfki.de/wiki/VoiceImportToolsTutorial

**Comment by nshmyrev** [comment]
@abitrolly latest wiki is on github:

https://github.com/marytts/marytts/wiki/VoiceImportToolsTutorial

that one on opendfki might be slightly outdated.

**Comment by abitrolly** [comment]
@nshmyrev thanks. Just need to get some free time now.

**Comment by maniyax** [comment]
@rugk Hello!
Sorry, but it's very bad quality.
Besides the text is missing in dataset.

I'm not sure about the German dataset, but russian dataset include a lot of different dictors.

**Comment by beqabeqa473** [comment]
hello.

common voice as i know is targetted for creation of rnn voices.
the way they're built is a bit different. this method doesn't
specifically need one voice as it builds acoustic model. but for
rhvoice it is a requirement of good quality recording of one speaker
to create a voice as technologies are different.

On 10/17/20, Artem Plaksin <notifications@github.com> wrote:
> @rugk Hello!
> Sorry, but it's very bad quality.
> Besides the text is missing in dataset.
>
> I'm not sure about the German dataset, but russian dataset include a lot of
> different dictors.
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/Olga-Yakovleva/RHVoice/issues/24#issuecomment-710670992


-- 
with best regards beqa

**Comment by rugk** [comment]
Ah okay, thanks, this makes sense. Of course, yeah, their idea, was to collect many voices…

**Comment by winman3000** [comment]
Is there any progress on this in the meantime? Has anyone tried to import the voices from MaryTTS for RHVoice?

If there is no existing data we can use for this project, we will have to do everything from scratch. Unfortunately I have to repeat my questions:

1. How long does the audio need to be?
1. How exactly do you do the mapping if you have the text to go with it?
1. Is there really no material we could use for a German voice?

I would love to finish the project so that we finally have a German voice as well...

**Comment by citizenserious** [comment]
I have found some voices including German, I use them with VocalizerEx2 TTS. They are working fine, but I do not trust VocalizerEx2. Would be nice to have these voices in RHVoice (:

https://vocalizer-nvda.com/downloads

**Comment by zstanecic** [comment]
Not really! It will be the infringement of copyright, and it is not possible, because of the synthesis method.

**Comment by uncle-ben-devel** [comment]
Any progress?
I'd like to get involved as I very much like the voice output from this application and it's openness.
I am a german native speaker. How can I contribute voice / training data? I have decent recording equipment also.

**Comment by winman3000** [comment]
As far as I know, there is no progress on this. Apparently it is not enough to make language recordings, modules must also be adapted in C++, i.e. libraries must be developed for the German language.

**Comment by nshmyrev** [comment]
You need to ask Torsten https://github.com/thorstenMueller/Thorsten-Voice, he will make he voice for you.

**Comment by winman3000** [comment]
Maybe we should contact Torsten Müller and ask if he would be willing to port his voice for RHVoice. We will definitely need someone to write the modules for German in C++. Unfortunately I do not have programming skills myself.

**Comment by zstanecic** [comment]
Hi,

You don’t need to write anything in c++.

Please see, for example, the polish language as an example.

You can create the data only language, which is imported by RHVoice without knowing it by the engine.

You need to learn how foma language regexes work.

There are tutorials and regex references available.

**Comment by thorstenMueller** [comment]
Hi and sorry for joing late to discussion.

I tried to figure out required steps to add german language based on my Thorsten-Voice dataset. The "Polish" language was referenced but honestly i'm not sure what to do.
@zstanecic Would you mind helping me with the first steps?

**Comment by zstanecic** [comment]
Hi Thorsten,

First of all, you will need to define the language rules in the foma scripting language, then compile these to fst, with the foma compiler and interpreter. 

The version which is needed is v 0.9.18. 

Newer version causes issues.

Secondly, you will need to create the phonemes features in xml file, as defined for example for polish. For german you will need to modify thinks a bit, or probably from scratch.

Thirdly, you will need to have labelling.xml.

As the language you will create will be data only, you will need the language.conf and graph.tx.. This is for registering the letters and telling the synthesizer which of these are consonants, and which are wovels.

Feel free to ask questions if these will arise.

**Comment by thorstenMueller** [comment]
Hi @zstanecic ,

what i did so far:
- Installed `foma` in version `0.9.18`
- Created folder /data/languages/German
- In that "German" folder i've copied the following files from "Polish" directory: graph.txt, labelling.xml, language.conf, language.info, locale.info, phonemes.xml

Am i right, that ...
- i have to adjust all these files (or less or more) to german?
- running "foma" (do not know this tool at all, yet) will create several.fst and dt files?

Once this is done i guess i've to create a "Thorsten" folder in the /data/voices folder? And do whatever way of magic there ;-).

**Comment by zstanecic** [comment]
Hi Thorsten,

That’s not the right way to start.

You need to check the foma scripts first, and write the rules for your language.

Scripts are located in rhvoice/scripts/language.

You will need examples of data only languages.

Let me contact privately on some whatsapp or email to explain this further.

***@***.*** ***@***.***>

**Comment by thorstenMueller** [comment]
Thanks for your quick reply and support offer. As i've never worked with foma this is really highly appreciated.
Github mail doesn't show your mail please contact me here and we can talk by mail then.
https://www.thorsten-voice.de/en/contact/

**Comment by zstanecic** [comment]
Hi,

I sent you a msg. please check your inbox

**Comment by BluePixel4k** [comment]
@thorstenMueller thx for your effort. Maybe I can help a bit with adjusting the basic grammar and rules. But I don't know how exactly.

**Comment by svnpsc** [comment]
May this be helpful?
https://github.com/ikekonglp/NLPLAB/tree/master/FOMA_Scripts

**Comment by Foexle11** [comment]
any news?

**Comment by thorstenMueller** [comment]
Not from my side yet. Topic is absolutely interesting but remaining free time is (as mostly) is limited.

**Comment by citizenserious** [comment]
What about the [Piper ](https://github.com/rhasspy/piper) voices?

**Comment by thiswillbeyourgithub** [comment]
> What about the [Piper ](https://github.com/rhasspy/piper) voices?

Wondering that too

**Comment by winman3000** [comment]
Piper voices do not currently meet the usual fluent standard for me. RH Voice still sounds very synthetic, but the work with these voices is very good. Piper voices constantly change the melody of sentences because they are AI-based and depend on the context. Therefore, they are currently not so good when working with screen readers.

**Comment by thiswillbeyourgithub** [comment]
Well sure they'e far from perfect but there is no RH voice for french,  my native language. So the well trained (as there are different quality tiers) piper voices are to me totally usable and would be a great alternative to the only alternative: google's system.

Actually there's no project that I know of that could bring android tts for french, and that's probably the case for many other languages. So if this project is onboard to embed piper voices for unsupported languages  that would be a great service!

**Comment by BluePixel4k** [comment]
@thiswillbeyourgithub you can try Sherpa:
https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html
French is also available.

**Comment by thiswillbeyourgithub** [comment]
> @thiswillbeyourgithub you can try Sherpa: https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html French is also available.

My gawd I had never heard about this. Thank you so much

**Comment by rmcpantoja** [comment]
Hi,
Although sherpa has a lot of latency while using voices with it in screen readers, it is good, for example, to read books or so.
The main RHVoice's mission is the responsiveness, which is beneficial for us, screen reader users.

**Comment by abitrolly** [comment]
@PrivacyAndSecurity because this ticket stays open, and the repo is active, I believe there is no German language in RHVoice. The probability that RHVoice team will add it without any financial support is close to zero, so the best way is to just find the funding and hire them to research the language and add necessary model.

**Comment by abitrolly** [comment]
> But would they not accept if a German Person would help with this Voice?

I don't know. You need to ask @Olga-Yakovleva and/or other folks involved in the project directly. I think it is impossible to do the language without a German person involved, but even in this case and may still require a lot of time and dedication from the team to make it right. Remember that Olga too uses RHVoice to communicate online. It is a lot of magic for me in how it works.

> I just don't understand how all other Languages are available but not German yet? Because it was asked nearly 10 Years ago...It don't make much sense to me, because other Languages did get added in this 9 Years too...

Other languages were either native languages of developers, but I believe most of them just got funded with the help of governments that don't have own technology for its citizens. I believe both Belarus and Germany have own technology, or at least projects to develop such technology that is being distributed to "some people with families to feed", and that's why, even if we have these requests for decade, they are still not done.

**Comment by abitrolly** [comment]
@PrivacyAndSecurity voice data is not a problem. I believe there are already a lot of recorded voices. The problem is to program RHVoice to understand the structure of the language, its letters, how sentences are split into words and then pronounced, how different symbols in a sentence affect pronunciation. Things like this.

For Belarusian language we needed to gather corpus of text with stress marks to build "stress dictionary". Some people were doing it with governmental money and wanted to sell it, but there was no buyer, and I don't know where is the initiative now.

**Comment by domasofan** [comment]
Hi all,

Does someone know how far this project is going?
Would be great to have some activity.
What would be missing to get that into action?

Greetings and thanks,
Simon
